File: frontend/src/app/models/book.ts
export interface Author { id: number; name: string; bookCount: number; totalCopyCount: number; availableCopyCount: number; }
export interface Publisher { id: number; name: string; bookCount: number; totalCopyCount: number; availableCopyCount: number; }
export type Genre = 'KURGU' | 'KURGU_DISI' | 'BILIMKURGU' | 'FANTASTIK' | 'GIZEM' | 'TARIH' | 'BIYOGRAFI' | 'TEKNOLOJI';

export interface Book {
  id: number;
  title: string;
  isbn: string;
  publicationYear: number | null;
  summary: string | null;
  genre: Genre;
  coverImageUrl: string | null;
  createdAt: string;
  updatedAt: string;
  publisher: Publisher | null;
  authors: Author[];
  totalCopyCount: number;
  availableCopyCount: number;
  loanedCopyCount: number;
  overdueCopyCount: number;
}

export interface BookSummary { id: number; title: string; isbn: string; coverImageUrl: string | null; }
export interface BookRequest {
  title: string;
  isbn: string;
  publicationYear: number | null;
  summary: string | null;
  genre: Genre;
  coverImageUrl: string | null;
  publisherId: number | null;
  authorIds: number[];
}

File: backend/src/main/java/dev/yagiz/kulliyat/entity/Book.java
package dev.yagiz.kulliyat.entity;

import dev.yagiz.kulliyat.enums.Genre;
import jakarta.persistence.*;
import lombok.AccessLevel;
import lombok.Getter;
import lombok.Setter;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

@Entity
@Table(name = "books")
@Getter
@Setter
public class Book {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false)
    private String title;

    @Column(unique = true, nullable = false, length = 17)
    private String isbn;

    private Integer publicationYear;

    @Column(columnDefinition = "TEXT")
    private String summary;

    @Enumerated(EnumType.STRING)
    private Genre genre;

    @Column(name = "cover_image_url")
    private String coverImageUrl;

    @CreationTimestamp
    @Column(updatable = false)
    @Setter(AccessLevel.NONE)
    private LocalDateTime createdAt;

    @UpdateTimestamp
    @Setter(AccessLevel.NONE)
    private LocalDateTime updatedAt;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "publisher_id")
    private Publisher publisher;

    @ManyToMany(cascade = {CascadeType.PERSIST, CascadeType.MERGE})
    @JoinTable(
        name = "book_author",
        joinColumns = @JoinColumn(name = "book_id"),
        inverseJoinColumns = @JoinColumn(name = "author_id")
    )
    private List<Author> authors = new ArrayList<>();

    @OneToMany(mappedBy = "book", cascade = CascadeType.ALL, orphanRemoval = true)
    @Setter(AccessLevel.NONE)
    private List<BookCopy> copies = new ArrayList<>();

    public void addAuthor(Author author) {
        this.authors.add(author);
        author.getBooks().add(this);
    }

    public void removeAuthor(Author author) {
        this.authors.remove(author);
        author.getBooks().remove(this);
    }

}
